Controller/Callback/Offsite.php

```diff
@@ -12,6 +12,7 @@
 use Omise\Payment\Model\Config\Internetbanking;
 use Omise\Payment\Model\Config\Alipay;
 use Omise\Payment\Model\Config\Installment;
+use Omise\Payment\Model\Config\Truemoney;
 
 class Offsite extends Action
```

Comment:
After reading the code, it feels like this should maybe be in separate classes, rather than one Offsite class dealing with multiple different 'offsite' types

Reply:
could you please explain more why? what is the reason for that? 99% of the code would be the same.

Reply:
No, it wouldn't. You would have one parent class where all the stuff lives that is the same for every offsite method. Then you would have subclass for each particular offsite method where the unique could for that particualr offsite method would live. This keeps everything nicely separate, allows for easy changes (should any offsite method require something unusual), prevents the whole thing become monolithic  (full of ifs and switchblocks), and would allow for clean, easy addition/removal of new/outdated offsite methods

Reply:
Actually seeing this and I feel the same. This class is getting fat and packed.
Quick one that make this feels wrong is by updating this line
```php
if ($order->getState() !== Order::STATE_PENDING_PAYMENT) { }
```
to
```php
if (! in_array($order->getState(), [ Order::STATE_PENDING_PAYMENT, Order::STATE_PROCESSING ])) { }
```
Just updated one line but we have to worry if this will affect the rest methods. While separating would be much more simpler (and bonus point, cleaner).

_I might need to consider this one for WC as well thou_

Reply:
@guzzilar according to our long `cancellation problem` talks we should get here only `pending` state on live server. `true money` introduced also `processing`. What other edge cases do you have in mind?

Reply:
That is the problem, I don't know what it could possibly be as the one condition that used to applied for 3 payment methods was one thing, and now we altered that condition to make it support the 4th payment method.
Just raising a concern: "Will it affect or will it be any consequence for other payment methods as well?"

And that kind of concern normally happens when the code gets too fat (on the other hand, if we started to worry about this, then maybe our class is fat already).

Reply:
_Btw, above are just to back up the idea of why I also feel that this should maybe be in separate classes. Not saying I have any particular edge case in mind._

Reply:
ok, thanks @guzzilar @jonrandy .. refactoring that concerns other payment methods should be done in separate PR.

Reply:
👍 agreed

Reply:
I think this is bigger than just refactoring. It feels like sticking plaster code to make things work. It would be unclear to a new developer coming to this what is going on

Reply:
@jonrandy @guzzilar this is how it works in magento 2 project I am responsible for:
I come in.. I add a solution to an existing architecture.
If architecture needs to be changed there are 2 scenarios:
1) if it architecture change is urgent then new PR is opened to change that architecture. And then this old PR needs to apply new architecture before is merged.
2) new pr is done, merged, architecture is changed later.